### PR TITLE
nixos/garage: allow unsafe replication modes, restart on config change

### PR DESCRIPTION
**Pull request by lheckemann**
###### Description of changes
Allows using replication modes like `2-unstable` for _all the write speed_ (who needs read-after-write consistency?)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by jian-lin — APPROVED** [review]
LGTM

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
@ofborg test garage

File: nixos/modules/services/web-servers/garage.nix [hunk 1]
```diff
@@ -49,7 +49,7 @@ in
 
           replication_mode = mkOption {
             default = "none";
-            type = types.enum ([ "none" "1" "2" "3" 1 2 3 ]);
+            type = types.enum ([ "none" "1" "2" "3" "2-dangerous" "3-dangerous" 1 2 3 ]);
```
Comment (emilylange) [inline nixos/modules/services/web-servers/garage.nix:52]: While we are at it, is there any particular reason why `3-degraded` is missing from this `enum`? :)
Docs: https://garagehq.deuxfleurs.fr/documentation/reference-manual/configuration/#replication-mode

```suggestion
            type = types.enum ([ "none" "1" "2" "3" "2-dangerous" "3-dangerous" "3-degraded" 1 2 3 ]);
```

**Review by emilylange — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-servers/garage.nix (hunk 1 above)
Comment (lheckemann) [reply nixos/modules/services/web-servers/garage.nix:52]: Good shout!

**Review by emilylange — APPROVED** [review]

**Pull request merged**